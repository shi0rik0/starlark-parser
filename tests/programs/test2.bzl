def f(*args, **kwargs):
  return args, kwargs

def while_break(n):
  r = 0
  for i in range(1000):
    if n == 5:
      break
    r += n
    n -= 1
  return r

def while_continue(n):
  r = 0
  for i in range(1000):
    if n <= 0:
      break
    if n % 2 == 0:
      n -= 1
      continue
    r += n
    n -= 1
  return r

def fib(n):
  seq = []
  x = 0
  y = 1
  for i in range(n):
    if len(seq) == n:
      break
    seq.append(x)
    x, y = y, x + y
  return seq