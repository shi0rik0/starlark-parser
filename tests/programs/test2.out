[Def] f(*args,**kwargs)->{
    [Return] TUPLE(IDENTIFIER(args),IDENTIFIER(kwargs))
}
[Def] while_break(n)->{
    [Assign] IDENTIFIER(r)=INT(0)
    [For] IDENTIFIER(i)=CALL(IDENTIFIER(range),INT(1000)){
        [If] 
        EQ(IDENTIFIER(n),INT(5))->{
            [Break] 
        }
        [Assign] IDENTIFIER(r)+=IDENTIFIER(n)
        [Assign] IDENTIFIER(n)-=INT(1)
    }
    [Return] IDENTIFIER(r)
}
[Def] while_continue(n)->{
    [Assign] IDENTIFIER(r)=INT(0)
    [For] IDENTIFIER(i)=CALL(IDENTIFIER(range),INT(1000)){
        [If] 
        LE(IDENTIFIER(n),INT(0))->{
            [Break] 
        }
        [If] 
        EQ(MOD(IDENTIFIER(n),INT(2)),INT(0))->{
            [Assign] IDENTIFIER(n)-=INT(1)
            [Continue] 
        }
        [Assign] IDENTIFIER(r)+=IDENTIFIER(n)
        [Assign] IDENTIFIER(n)-=INT(1)
    }
    [Return] IDENTIFIER(r)
}
[Def] fib(n)->{
    [Assign] IDENTIFIER(seq)=LIST()
    [Assign] IDENTIFIER(x)=INT(0)
    [Assign] IDENTIFIER(y)=INT(1)
    [For] IDENTIFIER(i)=CALL(IDENTIFIER(range),IDENTIFIER(n)){
        [If] 
        EQ(CALL(IDENTIFIER(len),IDENTIFIER(seq)),IDENTIFIER(n))->{
            [Break] 
        }
        [Expr] CALL(DOT(IDENTIFIER(seq),append),IDENTIFIER(x))
        [Assign] TUPLE(IDENTIFIER(x),IDENTIFIER(y))=TUPLE(IDENTIFIER(y),ADD(IDENTIFIER(x),IDENTIFIER(y)))
    }
    [Return] IDENTIFIER(seq)
}
